pkg/valueset/pgx/postgresql_valueset.go

```diff
@@ -0,0 +1,151 @@
+package pgx
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/cloudprivacylabs/leap/pkg/utils"
+	"github.com/cloudprivacylabs/lsa/layers/cmd/valueset"
+	"github.com/jackc/pgx/v5"
+	"github.com/mitchellh/mapstructure"
+)
+
+func init() {
+	valueset.RegisterDB("pgx", NewPostgresqlDataStore)
+}
+
+type PostgesqlDataStore struct {
+	Database `mapstructure:",squash"`
+}
+
+func (pgx *PostgesqlDataStore) ValueSetLookup(ctx context.Context, tableId string, queryParams map[string]string) (map[string]string, error) {
+	return pgx.getResults(context.Background(), queryParams, tableId)
+}
+
+func (pgx *PostgesqlDataStore) GetTableIds() map[string]struct{} {
+	return pgx.tableIds
+}
+
+func (pgx *PostgesqlDataStore) Close() error {
+	return pgx.close()
+}
+
+func NewPostgresqlDataStore(value interface{}, env map[string]string) (valueset.ValuesetDB, error) {
+	psqlDs := &PostgesqlDataStore{}
+	if err := mapstructure.Decode(value, psqlDs); err != nil {
+		return psqlDs, err
+	}
+	psqlDs.tableIds = make(map[string]struct{})
+	for _, vs := range psqlDs.Valuesets {
+		psqlDs.tableIds[vs.TableId] = struct{}{}
+	}
+	psqlDs.URI = utils.ExpandVariables("${uri}", env)
+	psqlDs.User = utils.ExpandVariables("${pgx_user}", env)
+	psqlDs.Pwd = utils.ExpandVariables("${pwd}", env)
+	return psqlDs, nil
+}
+
+type Database struct {
+	DB           *sql.DB
+	DatabaseName string     `json:"db" yaml:"db"`
+	User         string     `json:"user" yaml:"user"`
+	Pwd          string     `json:"pwd" yaml:"pwd"`
+	URI          string     `json:"uri" yaml:"uri"`
+	Valuesets    []Valueset `json:"valuesets" yaml:"valuesets"`
+	tableIds     map[string]struct{}
+}
+
+type Valueset struct {
+	TableId string  `yaml:"tableId"`
+	Queries []Query `yaml:"queries"`
+}
+
+type Query struct {
+	Query string `yaml:"query"`
+}
+
+// setConnection sets the database connection at most once
+func (db *Database) setConnection() error {
```

Comment:
rename this to getConnection(), and return the connection from it, setting it if it was not set before. Use sync.Once to set it only once. Remove openDB, inline it here.